Improve debug info

fabiansperber opened the pull request:
Improve generation of llvm debug source locations:
- begin/end of procedure (e.g. previously stepping in the debugger would jump from end of proc to the beginning)
- fix scopes of some instruction in the for loop #1858

Embed a natvis file when building the compiler on windows as a debug build to improve debugging experience (at least in msvc) by providing views for String and Array<>

gingerBill commented:
The work looks really good so far, if this fixes a lot of the issues, A LOT OF PEOPLE will be very very happy.

JesseRMeyer commented:
This is a marked improvement for #1858.  However, the behavior is not perfect, as shown below.
![1](https://user-images.githubusercontent.com/5846309/178108632-2779ff6a-16c7-421a-b4e6-3debaff409b2.JPG)
![2](https://user-images.githubusercontent.com/5846309/178108636-8017326a-7ae3-4e29-ab37-976216993551.JPG)

`i`'s value tracking gets busted during the test condition check, but is restored on the assignment.

graphitemaster commented:
First of all, thank you @fabiansperber I'm very happy to see some progress being made on debug support. I think this is a very good start. I've pulled these changes and tested locally and can confirm this fixes some issues.

I'd like to extend a hand here. If there's anything you need help with or would like to combine forces and work on this together please don't hesitate to ask. I'd like to improve upon the debug information as well. I have some time off work to work on personal projects and improving debug support would greatly benefit me and it seems you have some LLVM experience.

Have you at all run into miscompilation issues with `-debug`. Those are primarily my focus right now, not exactly getting debug support to be 100% but just having it not break things would be huge itself. Unfortunately these changes don't help with that right now but that's okay, as these are already quite helpful on some smaller codebases I have written in Odin.

gingerBill commented:
I've implemented a lot of what you've done! (I did it since this was still a draft)

fabiansperber commented:
I've rebased my local branch with some more changes. The tracking of the loop variable should work now, @JesseRMeyer

Also fixed having all different variable-variants of a type-switch visible at the same time. Note that the typed variable is not available on the case statement itself, only on the first actual statement in that block - somehow llvm does still not like the debug info enough to do that
```
switch actual in a {
  case u64:   // actual is not available here :(
    fmt.println("u64") // actual is available here
  case string:
    fmt.println("string")
}
```

fabiansperber commented:
Also sorry for the late response, @graphitemaster, I did not have any miscompilations per se due to `-debug`, only the debug info not working correctly, e.g. being at stepping through wrong locations or "variables optimized out".

My little LLVM experience is only from looking at other compilers use LLVM and looking at the LLVM IR / comparing to clang's output. It's not too difficult to look at small examples in both odin and C and compare the IR output to see what's different and possibly wrong. Or as in some cases of this MR to just check if what is being generated, does make sense.

As for helping out, I see that there is a list of debug problems in #1877 which I will take a look at, when I have some free time. Also you can just ping me on discord as well ;)
